libyul/backends/evm/ControlFlowGraphBuilder.cpp

```diff
@@ -0,0 +1,467 @@
+			jump(afterLoop);
+	}
+	else
+	{
+		jump(loopCondition);
+		makeConditionalJump(std::visit(*this, *_loop.condition), loopBody, afterLoop);
+		m_currentBlock = &loopBody;
+		(*this)(_loop.body);
+		jump(post);
+		(*this)(_loop.post);
+		jump(loopCondition, true);
+	}
+
+	m_currentBlock = &afterLoop;
+}
+
+void ControlFlowGraphBuilder::operator()(Break const&)
+{
+	yulAssert(m_forLoopInfo.has_value(), "");
+	jump(m_forLoopInfo->afterLoop);
+	m_currentBlock = &m_graph.makeBlock();
+}
+
+void ControlFlowGraphBuilder::operator()(Continue const&)
+{
+	yulAssert(m_forLoopInfo.has_value(), "");
+	jump(m_forLoopInfo->post);
+	m_currentBlock = &m_graph.makeBlock();
+}
+
+void ControlFlowGraphBuilder::operator()(Leave const&)
+{
+	yulAssert(m_currentFunctionExit.has_value(), "");
+	m_currentBlock->exit = *m_currentFunctionExit;
+	m_currentBlock = &m_graph.makeBlock();
+}
+
+void ControlFlowGraphBuilder::operator()(FunctionDefinition const& _function)
+{
+	yulAssert(m_scope, "");
+	yulAssert(m_scope->identifiers.count(_function.name), "");
+	Scope::Function& function = std::get<Scope::Function>(m_scope->identifiers.at(_function.name));
+	m_graph.functions.emplace_back(&function);
+
+	yulAssert(m_info.scopes.at(&_function.body), "");
+	Scope* virtualFunctionScope = m_info.scopes.at(m_info.virtualBlocks.at(&_function).get()).get();
+	yulAssert(virtualFunctionScope, "");
+
+	auto&& [it, inserted] = m_graph.functionInfo.emplace(std::make_pair(&function, CFG::FunctionInfo{
+		_function.debugData,
+		function,
+		&m_graph.makeBlock(),
+		_function.parameters | ranges::views::transform([&](auto const& _param) {
+			return VariableSlot{
+				std::get<Scope::Variable>(virtualFunctionScope->identifiers.at(_param.name)),
+				_param.debugData
+			};
+		}) | ranges::to<vector>,
+		_function.returnVariables | ranges::views::transform([&](auto const& _retVar) {
+			return VariableSlot{
+				std::get<Scope::Variable>(virtualFunctionScope->identifiers.at(_retVar.name)),
+				_retVar.debugData
+			};
+		}) | ranges::to<vector>
+	}));
+	yulAssert(inserted, "");
+	CFG::FunctionInfo& info = it->second;
+
+	ControlFlowGraphBuilder builder{m_graph, m_info, m_dialect};
+	builder.m_currentFunctionExit = CFG::BasicBlock::FunctionReturn{&info};
+	builder.m_currentBlock = info.entry;
+	builder(_function.body);
+	builder.m_currentBlock->exit = CFG::BasicBlock::FunctionReturn{&info};
+}
+
+
+CFG::Operation& ControlFlowGraphBuilder::visitFunctionCall(FunctionCall const& _call)
+{
+	yulAssert(m_scope, "");
+	yulAssert(m_currentBlock, "");
+
+	if (BuiltinFunction const* builtin = m_dialect.builtin(_call.functionName.name))
+	{
+		Stack inputs;
+		for (auto&& [idx, arg]: _call.arguments | ranges::views::enumerate | ranges::views::reverse)
+			if (!builtin->literalArgument(idx).has_value())
+				inputs.emplace_back(std::visit(*this, arg));
+		CFG::BuiltinCall builtinCall{_call.debugData, *builtin, _call, inputs.size()};
+		return m_currentBlock->operations.emplace_back(CFG::Operation{
```

Comment:
Since this is `emplace`, I think you should specify constructor arguments directly here. Otherwise it runs the copy constructor so you could just as well do `push_back()`.
```suggestion
		return m_currentBlock->operations.emplace_back(
```

Reply:
Can't unless I add a constructor to the thing. In contrast to using ``push_back`` this *won't* copy, but move the operation and the compiler will probably be clever enough to eliminate the move here, since it's an implicit move constructor anyways (it's allowed to).

Reply:
Oh, fine then.